src/coordinate_transform/tie_points.rs

```diff
@@ -0,0 +1,355 @@
+        source_index: &OwnedRTree<f64>,
+        source_mesh: &Rc<Vec<Face>>,
+        target_mesh: &Rc<Vec<Face>>,
+        coord: &Coord,
+    ) -> Coord {
+        let index = source_index
+            .search(coord.x, coord.y, coord.x, coord.y)
+            .into_iter()
+            .find(|face_index| source_mesh[*face_index].contains(coord))
+            .unwrap();
+        let uv = source_mesh[index].locate(coord);
+        target_mesh[index].interpolate(uv)
+    }
+}
+
+#[derive(Debug)]
+pub struct Face {
+    boundary: Option<Boundary>,
+    support_points: [Coord; 3],
+}
+
+impl Face {
+    fn contains(&self, coord: &Coord) -> bool {
+        let Some(boundary) = &self.boundary else {
+            return true;
+        };
+
+        let check = |c1: &Coord, c2: &Coord| -> bool {
+            ((c2.x - c1.x) * (coord.y - c1.y) - (c2.y - c1.y) * (coord.x - c1.x)).is_sign_positive()
+        };
+
+        match boundary {
+            Boundary::Open {
+                coords,
+                from_direction,
+                to_direction,
+            } => {
+                check(&(coords[0] + *from_direction), &coords[1])
+                    && check(&coords[1], &(coords[1] + *to_direction))
+                    && coords.windows(2).all(|w| check(&w[0], &w[1]))
+            }
+            Boundary::Closed { coords } => {
+                let len = self.support_points.len();
+                (0..=len).all(|i| check(&coords[i], &coords[(i + 1) % len]))
+            }
+        }
+    }
+
+    fn locate(&self, coord: &Coord) -> [f64; 2] {
+        let [a, b, c] = &self.support_points;
+        let d = c.x * (a.y - b.y) - b.x * (a.y - c.y) + a.x * (b.y - c.y);
+        [
+            -(coord.x * (a.y - c.y) - c.x * (a.y - coord.y) + a.x * (c.y - coord.y)) / d,
+            (coord.x * (a.y - b.y) - b.x * (a.y - coord.y) + a.x * (b.y - coord.y)) / d,
+        ]
+    }
+
+    fn interpolate(&self, params: [f64; 2]) -> Coord {
+        let [u, v] = params;
+        let [a, b, c] = &self.support_points;
+        Coord {
+            x: -u * a.x - v * a.x + a.x + u * b.x + v * c.x,
+            y: -u * a.y - v * a.y + a.y + u * b.y + v * c.y,
+        }
+    }
+
+    fn compute_envelope(&self) -> (f64, f64, f64, f64) {
+        let Some(boundary) = &self.boundary else {
+            return (f64::MIN, f64::MIN, f64::MAX, f64::MAX);
+        };
+
+        let mut min_x = f64::MAX;
+        let mut min_y = f64::MAX;
+        let mut max_x = f64::MIN;
+        let mut max_y = f64::MIN;
+
+        let mut update_envelope = |coords: &[Coord]| {
+            for c in coords {
+                if c.x < min_x {
+                    min_x = c.x;
+                }
+                if c.x > min_x {
+                    min_x = c.x;
+                }
+                if c.y < min_x {
+                    min_x = c.y;
+                }
+                if c.y > min_x {
+                    min_x = c.y;
+                }
```

Comment:
Perhaps you meant this instead?

```rust
                if c.x < min_x {
                    min_x = c.x;
                }
                if c.x > max_x {
                    max_x = c.x;
                }
                if c.y < min_y {
                    min_y = c.y;
                }
                if c.y > max_y {
                    max_y = c.y;
                }
```

or even this?

```rust
                min_x = min_x.min(c.x);
                max_x = max_x.max(c.x);
                min_y = min_y.min(c.y);
                max_y = max_y.max(c.y);
```

Reply:
@gschulze, are you sure this is resolved? You're comparing both `c.x` and `c.y` to _only_ `min_x` in _all_ cases. Am I missing something?